### ISSUE TITLE
`crystal eval` or `crystal i`

### ISSUE THREAD

**Issue by beta-ziliani** [crystal-lang/crystal#10947]
Once #10910 is merged, we have to evaluate if it makes sense to have two commands for quickly executing some code.

Options are:
1. Keep everything, with the semantics of "today" (be that the day when `crystal i` is merged).
2. Depecrate `eval` in favor of `i`.
3. Make `eval` synonym of `i`.

I prefer 2 above but, most of all, for the moment I just want to attract this discussion to its own issue instead discussing it in #10910.

**Comment by beta-ziliani** [comment]
@Sija I didn't understand your comment:

> I'm not sure that's going to be a worthy change, since `eval` has its purpose for one liners.

I understand that `crystal i` will also have the option of executing one line and without all the burden of compilation. That's what you mean?

**Comment by wontruefree** [comment]
I think there should be an option 4 which is not to make an `i` command and just use `eval`.

1. `crystal eval` starts a REPL and run why @asterite is working on.
2. `crystal run` would be able to take one liners and run them

`run` would handle compiling and running blocks of text either from a file, string, or stdin
`eval` would basically be a REPL and it would also be able to take input from STDIN kind of like it does now but better.

**Comment by vlazar** [comment]
I think `crystal i` sounds cool (crystal eye). Well done @asterite

**Comment by asterite** [comment]
I really don't think we should change anything. Right now `crystal i` can take arguments which become ARGV. If you pass stdin to `crystal i`, that content should be in STDIN inside the program, it should not become the program to run.

It's also hard to make i run one liners like that.

**Comment by HertzDevil** [comment]
If the program source comes from the standard input, the standard way would be using `-` in place of normal file names, whether it's `run`, `build` or `i`. So:

```crystal
$ echo 'puts 1 + ARGV.first.to_i' | crystal run - -- 2
3

$ echo 'puts 1 + ARGV.first.to_i' | crystal build -o a.out -
$ ./a.out 2
3

$ echo 'puts 1 + ARGV.first.to_i' | crystal i - -- 2
3

$
```

With this, `crystal eval` without arguments becomes equivalent to `crystal run -`. Likewise, extra source code coming from command line arguments should require an option like `-e`:

```crystal
$ crystal run -e 'puts 1 + ARGV.first.to_i' -- 2
3

$ crystal build -o a.out -e 'puts 1 + ARGV.first.to_i'
$ ./a.out 2
3

$ crystal i -e 'puts 1 + ARGV.first.to_i' -- 2
3

$
```

Then we could deprecate `eval` without changing the behaviour of `run`, `build`, or `i`.

**Comment by Sija** [comment]
@beta-ziliani I was referring to the clear separation of each of the commands. You don't need to dig into details when you think of using `eval`, `run`, or `i` (interactive). Otherwise reading the manual, `--help` or other resources becomes pretty much a must to understand the differences.

**Comment by jhass** [comment]
I definitely have `crystal eval` from standard input in use and I'm not sure `crystal i` is a good replacement for that usecase, at least just yet. You can probably guess where I use it :P

**Comment by HertzDevil** [comment]
`crystal eval` from standard input is actually equivalent to `crystal run --stdin-filename eval`, by the way.

**Comment by HertzDevil** [comment]
Another thing I discovered is that it is entirely possible for valid Crystal source code to start with one of `crystal eval`'s option flags:

```
$ crystal eval "-Dir.current.size <= 0 ? p 1 : p 2"
{% p flag?(:"ir.current.size > 0 ? p 1 : p 2") %}^D
true
```

`-D` defines a compile-time flag, so the whole program is parsed as a command-line option, and `crystal eval` asks for standard input and runs that instead, thus merely quoting the entire program is insufficient. `crystal run -e` would eliminate such surprises.
